README.md
Log\Beffered
========

## NAME

Log\Buffered -

## SYNOPSIS

```php
use Log\Buffered;

require_once '/path/to/vendor/autoload.php';

$log = new Buffered(
    [
        'file'        => '/paht/to/myapp.log', // stderr by default
        'buffer_size' => 10240,                // 5120 bytes by default
    ]
);
$log->debugf('This is a %s message', 'debug'); // This message never be output
$log->infof('This is an %s message', 'info');
$log->warnf('This is a %s message', 'warn');
$log->critf('This is a %s message', 'crit');
```

## INSTALLATION
To install this package into your project via composer, add the following snippet to your `composer.json`. Then run `composer install`.

```
"require": {
    "travail/log-buffered": "dev-master"
}
```

If you want to install from gihub, add the following:

```
"repositories": [
    {
        "type": "vcs",
        "url": "git@github.com:travail/php-Log-Buffered.git"
    }
]
```

## DEPENDENCIES

Log\Buffered has the dependency on the following:

* [Log\Minimal](https://github.com/travail/php-Log-Minimal)

## METHODS

### __construct

`__constract(array $attrs)`

#### Parameters

`$attrs` can contain the keys below:

##### file

Path to a file into which logger writes.

##### color

Outputs colored messages if set a true value, `false` by default.

##### debug

Outputs debug messages if set a true value, `false` by default.

##### log_level

The lower limit of the log level. `info` by default.
**NOTE: To output debug messages set `debug` to `log_level` and a true value to `debug`.**

##### trace_level

The depth of stack trace. `1` by default.

##### buffer_size

Set the size of buffer which allow to buffer messages within given size. `5120 bytes` by default. You must specify this between `Log\Buffered::MIN_BUFFER_SIZE` and `Log\Buffered::MAX_BUFFER_SIZE`, or an `Exception` will be thrown.

### debugf

`void debugf(string $format [, mixed $val [, mixed ...]])`

### infof

`void infof(string $format [, mixed $val [, mixed ...]])`

### warnf

`void infof(string $format [, mixed $val [, mixed ...]])`

### critf

`void infof(string $format [, mixed $val [, mixed ...]])`

These methods above buffer messages until buffered messages reach `buffer_size` you specified. If you try to buffer messages more than `buffer_size` the logger will flush buffered messages automatically with the caution below.

```
***************************** CAUTION!!! **********************************
Flush the buffer, because the size of buffered messages reached 1024.
If you want to buffer more messages, set the more large size to buffer_size
in the constructor or by calling Log\Buffered::setBufferSize().
```

#### Parameters

The same as the built-in function `sprintf`.

### append

`void append(string $message)`

Append a given message to the buffer. Usually you don't have to use this directly.

#### Parameters

##### $message

A string to be buffered.

### flush

`void flush(void)`

Flush buffered messages.

### clear

`void clear(void)`

Clear the buffer.

### getMinBufferSize

`int getMinBufferSize(void)`

Returns the size of min buffer size `1024` bytes.

### getMaxBufferSize

`int getMaxBufferSize(void)`

Returns the size of max buffer size `5120000` bytes.

### getBufferedSize

`int getBufferedSize(void)`

Returns the size of buffered messages.

## ENABLING DEBUG MESSAGES

There are two ways to enable debug messages:

1. By Environment Values
1. By Constructor

### By Environment Values

```php
$_SERVER['LM_DEBUG']     = true;
$_SERVER['LM_LOG_LEVEL'] = 'debug';
$log = new Log\Buffered();
$log->debugf('This is a %s message: %s', 'debug');
```

### By Constructor

```php
$log = new Log\Buffered(
    [
        'debug'     => true,
        'log_level' => 'debug',
    ]
);
$log->debugf('This is a %s message: %s', 'debug');
```

## AUTHOR

travail

## LICENSE

This library is free software. You can redistribute it and/or modify it under the same terms as PHP itself.


CHANGELOG.md
CHANGELOG
========

## 0.2.0 2020-08-31

* Fix the deprecated feature
  * Passing null to non-nullable parameters of built-in functions
* Fix the document

## 0.1.0 2014-05-01

* Composerized
* Tidy up README.md and example scripts to compliant with composer


lib/Buffered.php
<?php

namespace Log;

use InvalidArgumentException;

class Buffered extends Minimal
{
    /**
     * @var string The version of this package
     */
    public const VERSION = '0.2.0';

    /**
     * @var int Min buffer size
     */
    public const MIN_BUFFER_SIZE = 1024; // 1K

    /**
     * @var int Max buffer size
     */
    public const MAX_BUFFER_SIZE = 5120000; // 5M

    /**
     *  @var int Default buffer size
     */
    public const DEFAULT_BUFFER_SIZE = 5120; // 5K

    /**
     * @var string Default file path to which logger write
     */
    public const DEFAULT_FILE = 'php://stderr';

    /**
     * @var string Buffer in which logger store messages
     */
    protected $buffer= '';

    /**
     * @var int Buffer size
     */
    protected $buffer_size = self::DEFAULT_BUFFER_SIZE;

    /**
     * @var string File path to which logger write, self::DEFAULT_FILE by default
     */
    protected $file = self::DEFAULT_FILE;

    /**
     * @var resource File handle
     */
    protected $fd;

    /**
     * Create a new instance of Log\Buffered
     *
     * @param array $attrs
     * @throws InvalidArgumentException
     */
    public function __construct($attrs = [])
    {
        if (isset($attrs['color'])) {
            parent::$color = $attrs['color'];
        }
        if (isset($attrs['debug'])) {
            parent::$debug = $attrs['debug'];
        }
        if (isset($attrs['log_level'])) {
            parent::$log_level = $attrs['log_level'];
        }
        if (isset($attrs['trace_level'])) {
            parent::$trace_level = $attrs['trace_level'];
        }

        if (isset($attrs['file']) && file_exists($attrs['file'])) {
            if (!is_readable($attrs['file'])) {
                throw new InvalidArgumentException(sprintf('%s is not readable', $attrs['file']));
            }
            if (!is_writable($attrs['file'])) {
                throw new InvalidArgumentException(sprintf('%s is not writable', $attrs['file']));
            }
        }

        if (isset($attrs['file'])) {
            $this->file = $attrs['file'];
        }
        $this->fd = fopen($this->file, 'a+');

        if (isset($attrs['buffer_size']) && $attrs['buffer_size'] < self::MIN_BUFFER_SIZE) {
            throw new InvalidArgumentException('buffer_size must be more than ' . self::MIN_BUFFER_SIZE);
        }
        if (isset($attrs['buffer_size']) && $attrs['buffer_size'] > self::MAX_BUFFER_SIZE) {
            throw new InvalidArgumentException('buffer_size must be less than ' . self::MAX_BUFFER_SIZE);
        }
        if (isset($attrs['buffer_size'])) {
            $this->buffer_size = $attrs['buffer_size'];
        }

        $obj = $this;
        parent::$print = static function($time, $type, $msg, $trace, $raw_msg) use ($obj)
        {
            $obj->append(sprintf("%s [%s] %s at %s line %s\n",
                    $time, $type, $msg, $trace['file'], $trace['line']));
        };
    }

    /**
     * Append a message into the buffer.
     *
     * @param string $message A message
     * @return void
     */
    public function append($message): void
    {
        $this->buffer .= $message;
        if ($this->getBufferedSize() > $this->buffer_size) {
            $this->overflow();
        }
    }

    /**
     * Flush messages.
     *
     * @return void
     */
    public function flush(): void
    {
        fwrite($this->fd, $this->buffer);
        $this->clear();
    }

    /**
     * Clear the buffer.
     *
     * @return void
     */
    public function clear(): void
    { 
        $this->buffer = '';
    }

    /**
     * Print the caution and flush the buffer
     * when buffered messages reach Log\Buffered::MAX_BUFFER_SIZE.
     *
     * @return void
     */
    protected function overflow(): void
    {
        $this->buffer .= sprintf(
            "***************************** CAUTION!!! **********************************\n" .
            "Flush the buffer, because the size of buffered messages reached %d.\n" .
            "If you want to buffer more messages, set the more large size to buffer_size\n" .
            "in the constructor or by calling Log\Buffered::setBufferSize().\n" .
            "***************************************************************************\n",
            $this->buffer_size);
        $this->flush();
    }

    /**
     * Return Log\Buffered::MIN_BUFFER_SIZE.
     *
     * @return int
     */
    public function getMinBufferSize(): int
    {
        return self::MIN_BUFFER_SIZE;
    }

    /**
     * Return Log\Buffered::MAX_BUFFER_SIZE.
     *
     * @return int
     */
    public function getMaxBufferSize(): int
    {
        return self::MAX_BUFFER_SIZE;
    }

    /**
     * Return the size of buffer.
     *
     * @return int Size of buffer
     */
    public function getBufferSize(): int
    {
        return $this->buffer_size;
    }

    /**
     * Set passed size to buffer size.
     *
     * @param int $size Size of the buffer
     * @return void
     */
    public function setBufferSize($size): void
    {
        $this->buffer_size = $size;
    }

    /**
     * Return current size of buffered messages.
     *
     * @return int Current size of buffered messages
     */
    public function getBufferedSize(): int
    {
        if ($this->buffer === null) {
            return 0;
        }
        return strlen($this->buffer);
    }

    /**
     * Return the path to file to which logger print messages.
     *
     * @return string The path to file to which logger print messages
     */
    public function getFile(): string
    {
        return $this->file;
    }

    /**
     * Flush the buffer when messages buffered.
     *
     * @return void
     */
    public function __destruct()
    {
        if ($this->getBufferedSize() > 0) {
            $this->flush();
        }
    }
}
